src/kubernetes/client/config/mod.rs

```diff
@@ -0,0 +1,27 @@
+//! Client configuration.
+
+use crate::tls::TlsOptions;
+use http02::Uri;
+
+/// A k8s client configuration.
+///
+/// This type is designed to hold all possible variants of the configuration.
+/// It also abstracts the client from the various ways to obtain the
+/// configuration.
+///
+/// The implementation is fairly limited, and only covers the use cases we
+/// support.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The base URL to use when constructing HTTP requests to the k8s API
+    /// server.
+    pub base: Uri,
+
+    /// The bearer token to use at the `Authorization` header.
+    pub token: String,
+
+    /// The TLS configuration parameters to use at the HTTP client.
+    pub tls_options: TlsOptions,
+}
+
+pub mod in_cluster;
```

Comment:
Typically mod statements happen just below the `use` statements, I suspect `cargo fmt` will mop this up for you. :)

Reply:
`cargo fmt` allows it :)
The semantics here is that `in_cluster` builds upon the `Config` type - thus it's defined afterward. There's no language requirement to do this, I just think it adds a bit of extra readability.
I don't mind moving it upper though if it didn't help with readability - after all, maybe having mods all the way up helps better. :smile:

Reply:
> It is extra readable if you know that it builds upon the Config file, without that, it looks to me out of place.

This is two-fold. :) Do you have an opinion on moving it to the upper part of the file?

---

In general, for every syntax/styling related question, I'd suggest that we don't bother even discussing those. Not just in this PR, but in general. The reason is it often touches subjective preferences over what's good and what's bad, thus making it impossible to achieve an optimal solution. Furthermore, it's usually doesn't make any practical difference. Additionally, we have a code autoformatter for the most important styling rules, which is the only practically working way to enforce certain styling rules reliably.

I'm not complaining, and I'm not proposing to mute this discussion in the current PR, on the contrary - I'm grateful for the comment and the opportunity to discuss and learn about the perception of the code here. In general, though, I personally try to not request changes to styling details when reviewing PRs.

Reply:
Ok, then let's try to sort this out without involving subjective preferences. Of course that is generally not possible in any case since I'll have to use some principles. As an extra I'll try to keep it short.

I argue that KISS applies here. Example. the extra meaning 
>in_cluster builds upon the Config type - thus it's defined afterward

is fragile. It's so easy for other contributors to not notice that the meaning exists, and therefore easy to be broken either by changing the content of the mod or by adding additional mods that build upon  Config type.

Reply:
It's good a point. I.e. here we build upon the fact that `in_cluster` build upon the `Config`, but things can be moved around or added in such a way that it's no longer as elegant, so it's better to eliminate this possibility now than do an extra commit later.
I'll move it to the top! :+1:

Reply:
Thanks for moving it Mike, I regret that so much discussion had to happen.

> In general, for every syntax/styling related question, I'd suggest that we don't bother even discussing those. Not just in this PR, but in general. The reason is it often touches subjective preferences over what's good and what's bad, thus making it impossible to achieve an optimal solution. Furthermore, it's usually doesn't make any practical difference. Additionally, we have a code autoformatter for the most important styling rules, which is the only practically working way to enforce certain styling rules reliably.
> 
> I'm not complaining, and I'm not proposing to mute this discussion in the current PR, on the contrary - I'm grateful for the comment and the opportunity to discuss and learn about the perception of the code here. In general, though, I personally try to not request changes to styling details when reviewing PRs.

I don't think we should discuss them either. I think if one of your teammates takes the effort to ask you to make a minor unimportant styling change you should just do it, without needing to require several comments and a discussion. I pointed it out because I thought you missed it, and only you to move a mod statement to where mod statements, by convention, go. Not change logic or do something controversial.

Reply:
Oh, I see. This is largely my fault, I misunderstood the intent. But then, I could've hypothetically, either just decline the change request or implement it, without thinking, and having this whole discussion. But then, I wouldn't have learned and would likey to pull the same next time. :smile: Now I understand the pitfalls I didn't think about, so it's clearly beneficial in the long run. I'm very satisfied with how it ended up, and I think we did everything right.

Reply:
Me too! :)